tests/func/experiments/test_experiments.py

```diff
@@ -107,67 +107,72 @@ def test_failed_exp_workspace(
         [
-            ["goo.bag=4"],
-            "{foo: [bar: 1, baz: 2], goo: {bag: 4}, lorem: false}",
+            ["foo.0=bar"],
+            "foo:\n- bar\n- baz: 2\ngoo:\n  bag: 3.0\nlorem: false",
         ],
-        [["foo[0]=bar"], "{foo: [bar, baz: 2], goo: {bag: 3}, lorem: false}"],
         [
-            ["foo[1].baz=3"],
-            "{foo: [bar: 1, baz: 3], goo: {bag: 3}, lorem: false}",
+            ["foo.1.baz=3"],
+            "foo:\n- bar: 1\n- baz: 3\ngoo:\n  bag: 3.0\nlorem: false",
         ],
         [
-            ["foo[1]=[baz, goo]"],
-            "{foo: [bar: 1, [baz, goo]], goo: {bag: 3}, lorem: false}",
+            ["goo.bag=4.0"],
+            "foo:\n- bar: 1\n- baz: 2\ngoo:\n  bag: 4.0\nlorem: false",
         ],
+        [
+            ["params.yaml:++goo={bag: 1, b: 2}"],
+            "foo:\n- bar: 1\n- baz: 2\ngoo:\n  bag: 1\n  b: 2\nlorem: false",
+        ],
+        [
+            ["++goo={bag: 1, b: 2}"],
+            "foo:\n- bar: 1\n- baz: 2\ngoo:\n  bag: 1\n  b: 2\nlorem: false",
+        ],
+        # 6129
+        [
+            ["foo="],
+            "foo: ''\ngoo:\n  bag: 3.0\nlorem: false",
+        ],
+        # 6129
+        [
+            ["foo=null"],
+            "foo:\ngoo:\n  bag: 3.0\nlorem: false",
+        ],
+        # 5868
+        [
+            ["foo=1992-11-20"],
+            "foo: '1992-11-20'\ngoo:\n  bag: 3.0\nlorem: false",
+        ],
+        # 5868
+        [
+            ["foo='1992-11-20'"],
+            "foo: '1992-11-20'\ngoo:\n  bag: 3.0\nlorem: false",
+        ],
+        # Appending
+        [
+            ["+a=1"],
+            "foo:\n- bar: 1\n- baz: 2\ngoo:\n  bag: 3.0\nlorem: false\na: 1",
+        ],
+        # Removing
+        [["~foo"], "goo:\n  bag: 3.0\nlorem: false"],
     ],
 )
-def test_modify_params(tmp_dir, scm, dvc, mocker, changes, expected):
-    tmp_dir.gen("copy.py", COPY_SCRIPT)
-    tmp_dir.gen(
-        "params.yaml", "{foo: [bar: 1, baz: 2], goo: {bag: 3}, lorem: false}"
-    )
-    stage = dvc.run(
-        cmd="python copy.py params.yaml metrics.yaml",
-        metrics_no_cache=["metrics.yaml"],
-        params=["foo", "goo", "lorem"],
-        name="copy-file",
-    )
-    scm.add(["dvc.yaml", "dvc.lock", "copy.py", "params.yaml", "metrics.yaml"])
-    scm.commit("init")
-
-    new_mock = mocker.spy(dvc.experiments, "new")
-    results = dvc.experiments.run(stage.addressing, params=changes)
-    exp = first(results)
-
-    new_mock.assert_called_once()
-    fs = scm.get_fs(exp)
-    with fs.open("metrics.yaml", mode="r") as fobj:
+def test_modify_params(params_repo, dvc, changes, expected):
+    dvc.experiments.run(params=changes)
+    # pylint: disable=unspecified-encoding
+    with open("params.yaml", mode="r") as fobj:
         assert fobj.read().strip() == expected
 
 
 @pytest.mark.parametrize(
     "changes",
-    [["lorem.ipsum=3"], ["foo[0].bazar=3"]],
+    [["foobar=2"], ["lorem=3,2"], ["+lorem=3"], ["foo[0]=bar"]],
```

Comment:
> If I so `dvc exp run --set-param newparam=foo` with a parameter that doesn't already exist, this fails in `main` but succeeds here. I think it should only succeed if using `dvc exp run --set-param +newparam=foo`.

It should work as expected now. Unexisting params should raise an exception unless `+`/`++` is passed